src/util/prolog/prolog.test.ts

```diff
@@ -0,0 +1,246 @@
+				const result = await session.executeQuery("likes(lisa, X, Y, Z).");
+
+				expect(result).toEqual([["bob", "karl", "friedrich"], ["bob", "friedrich", "karl"], ["false"]]);
+
+		});
+
+		test("Four Variables Two Results", async () => {
+				const session = new PrologSession(`
+					likes(lisa, bob, karl, friedrich, dieter).
+					likes(lisa, bob, karl, dieter, friedrich).
+					`, 1000);
+				const result = await session.executeQuery("likes(lisa, X, Y, Z, A).");
+
+				expect(result).toEqual([["bob", "karl", "friedrich", "dieter"], ["bob", "karl", "dieter", "friedrich"], ["false"]]);
+
+		});
+
+		test("Cut Operator Maximum Function", async () => {
+
+				const session = new PrologSession("max(X,Y,Y)  :-  X  =<  Y, !. max(X, Y, X).", 1000);
+				const result = await session.executeQuery("max(3, 7, X).");
+				expect(result).toEqual([["7"], ["false"]]);
+
+		});
+
+		test("Cut Operator Endless Loop", async () => {
+
+				const session = new PrologSession("likes(lisa, bob, karl). likes(X, Y, Z):- likes(X, Z, Y), !.", 1000);
+				const result = await session.executeQuery("likes(lisa, Y, Z).");
+				expect(result).toEqual([["bob", "karl"], ["karl", "bob"], ["false"]]);
+
+		});
+
+		test("Only Write", async () => {
+
+				const session = new PrologSession("write('Hello').", 1000);
+				const result = await session.executeQuery("");
+				expect(result).toEqual([["false"]]);
+
+		});
+
+		test("Write Function and Query", async () => {
+
+				const session = new PrologSession("write('Hello'). likes(lisa, bob).", 1000);
+				const result = await session.executeQuery("likes(lisa, bob).");
+				expect(result).toEqual([["true"], ["false"]]);
+
+		});
+
+		test("Truth Value (True) Perfomance", async () => {
+				const session = new PrologSession("likes(bob, lisa).", 1000);
+				const result = await session.executeQuery("likes(bob, lisa).");
+
+				expect(result).toEqual([["true"], ["false"]]);
+		});
+});
+
+	describe("Family Program", () => {
+		// To make it more clear what the code is supposed to give back the family title is used
+		// instead of names
+		// grandfatherf for grandfather father side of the family
+		// grandfatherm for grandfather mother side of the family
+		const session = new PrologSession(`
+			male(dad).
+			male(son).
+			male(grandfatherm).
+			male(grandfatherf).
+
+			female(mom).
+			female(daugther).
+			female(grandmotherm).
+			female(grandmotherf).
+
+			parent(dad, daugther).
+			parent(mom, daugther).
+			parent(dad, son).
+			parent(mom, son).
+			parent(grandfatherm, mom).
+			parent(grandfatherf, dad).
+			parent(grandmotherm, mom).
+			parent(grandmotherf, dad).
+
+			father(X, Y):-
+			parent(X, Y),
+			male(X).
+
+			mother(X, Y):-
+			parent(X, Y),
+			female(X).
+
+			grandparent(X, Z):-
+			parent(X, Y),
+			parent(Y, Z).
+
+			grandfather(X, Z):-
+			grandparent(X, Z),
+			male(X).
+
+			grandmother(X, Z):-
+			grandparent(X, Z),
+			female(X).
+				`, 1000);
+
+		test.each`
```

Comment:
if you use `test.each`, please make sure you format the text so it actually increases readability.

Reply:
I think it is more readibly since it get rid of the repeated structurs which only clutter the code. I can at least find my input and expected output easier this way. It also looks different in the editor. The comments make it less readable on github.

This also has another advantage. If I change the way the results are accessed I only have to change it one location for all tests and not in every test method.

Reply:
The indentation is a bit weird, maybe you could move the inputs one more level to the left?